src/ui/marker.js

```diff
@@ -346,4 +362,116 @@ export default class Marker {
+            this._state = 'active';
+
+            /**
+             * Fired when dragging starts
+             *
+             * @event dragstart
+             * @memberof Marker
+             * @instance
+             * @type {Object}
+             * @property {Marker} marker object that is being dragged
+             */
+            this.fire(new Event('dragstart'));
+        }
+
+        /**
+         * Fired while dragging
+         *
+         * @event drag
+         * @memberof Marker
+         * @instance
+         * @type {Object}
+         * @property {Marker} marker object that is being dragged
+         */
+        this.fire(new Event('drag'));
+    }
+
+    _onUp() {
+        // revert to normal pointer event handling
+        this._element.style.pointerEvents = 'auto';
+        this._positionDelta = null;
+        this._map.off('mousemove', this._onMove);
+        this._map.off('touchmove', this._onMove);
+
+        // only fire dragend if it was preceded by at least one drag event
+        if (this._state === 'active') {
+            /**
+            * Fired when the marker is finished being dragged
+            *
+            * @event dragend
+            * @memberof Marker
+            * @instance
+            * @type {Object}
+            * @property {Marker} marker object that was dragged
+            */
+            this.fire(new Event('dragend'));
+        }
+
+        this._state = 'inactive';
+    }
+
+    _addDragHandler(e: MapMouseEvent | MapTouchEvent) {
+        if (this._element.contains((e.originalEvent.target: any))) {
+            e.preventDefault();
+
+            // We need to calculate the pixel distance between the click point
+            // and the marker position, with the offset accounted for. Then we
+            // can subtract this distance from the mousemove event's position
+            // to calculate the new marker position.
+            // If we don't do this, the marker 'jumps' to the click position
+            // creating a jarring UX effect.
+            this._positionDelta = e.point.sub(this._pos).add(this._offset);
+
+            this._state = 'pending';
+            this._map.on('mousemove', this._onMove);
+            this._map.on('touchmove', this._onMove);
+            this._map.once('mouseup', this._onUp);
+            this._map.once('touchend', this._onUp);
+        }
+    }
+
+    /**
+     * Sets the `draggable` property and functionality of the marker
+     * @param {boolean} [shouldBeDraggable=false] Turns drag functionality on/off
+     * @returns {Marker} `this`
+     */
+    setDraggable(shouldBeDraggable: boolean) {
+        this._draggable = !!shouldBeDraggable; // convert possible undefined value to false
+
+        // handle case where map may not exist yet
+        // e.g. when setDraggable is called before addTo
+        if (this._map) {
+            if (shouldBeDraggable) {
+                this._map.on('mousedown', this._addDragHandler);
```

Comment:
We'll need to either fix #4597 or check that `this._draggable` state is actually changing, so as not to register duplicate handlers if `setDraggable(true)` is called when the marker is already draggable.

Reply:
a5da36c addresses #4597 which would resolve this comment. Is that an acceptable solution? It definitely works.